wip

alexcrichton opened the pull request.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+still plenty of interface types to benefit in the meantime. This canonical ABI
+is intended to be a stepping stone for interface types to advance, and in the
+future with the full adapter language and adapter fusion the canonical ABI can
+be simply expressed in terms of this adapter language.
```
```suggestion
This document proposes a "canonical ABI" that maps interface-typed function
signatures down to core wasm function signatures that can be imported or
exported by core wasm, without the need for custom adapter functions.
While the end goal of interface types is to divorce interface signatures from
the actual underlying representation of each type through custom adapter
functions, there's still a decent amount of remaining work to stabilize adapter 
functions and so one goal of defining a "canonical ABI" is to allow us to define
an interface types MVP that can be released in the short term. In addition to 
serving as a stepping stone, the canonical ABI will also be valuable in the long 
term in providing a simple path for languages' tooling to adopt interface types 
by binding to the canonical ABI like any normal C API and then using shared
language-agnostic tooling to wrap the core module into an interface-typed 
adapter module.
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+lowering depending on the context of an imported or exported function. The four
+operations defined are:
+
+* Lowering an imported function call - this means that a WebAssembly module is
```
I think perhaps it would be a bit more consistent with interface types terminology if you could reframe these 4 bullets as: lifting an import's params, lowering an import's results, lowering an export's params, lifting an export's results

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+   generator for the appropriate language will be used to have an API, in that
+   language, which takes types of that language (e.g. Rust structs, C unions,
+   etc) and then converts those values to the canonical ABI (according to this
+   document). The WebAssembly will have a core wasm import of a WASI function
+   which is expected to behave according to this document.
```
```suggestion
   document). The WebAssembly module will have a core wasm import of a WASI function
   whose core signature and behavior is defined by this document.
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+   document). The WebAssembly will have a core wasm import of a WASI function
+   which is expected to behave according to this document.
+
+3. On the implementor side of things, a JS host, for example, would use a code
```
```suggestion
3. On the host side of things, a JS host, for example, would use a code
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+   which is expected to behave according to this document.
+
+3. On the implementor side of things, a JS host, for example, would use a code
+   generator to generate an object where it takes core wasm values and then
+   hooks up to a JS implementation of WASI. The JS implementation would only use
+   JS objects and wouldn't need to interact with the wasm module/instance itself
+   (due to the shared-nothing nature of interface types).
```
```suggestion
   generator to generate imported JS glue functions which take and return core wasm values
   and, converting these to and from JS values, call the JS implementation of WASI.
   The JS WASI implementation would thus only deal with JS values and wouldn't need to interact 
   with core wasm values or linear memory (due to the shared-nothing nature of interface types).
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+   JS objects and wouldn't need to interact with the wasm module/instance itself
+   (due to the shared-nothing nature of interface types).
+
+Eventually with a binary encoding of interface types it can be imagined that
+WebAssembly modules are wrapped in a thin binary encoding which ascribes the
+types of all imports and exports. With these types various interface types
+signatures can be checked at instantiation time to ensure that the stars all
+align and the ABIs are expected to match. In the very immediate future, however,
+a binary encoding is optional because it can simply be understood that a wasm
+module's imports/exports are described with interface types and both sides of
+the boundary are using the same interface types to generate glue code.
```
I'd leave out this paragraph.  While *this* doc doesn't define the binary encoding, another will, so it's perhaps better just to leave it open in this doc how exactly these types "show up" in a module/component.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+            | (list <intertype>)
+            | (record (field <name> <id>? <intertype>)*)
+            | (variant (case <name> <id>? <intertype>?)*)
+            | (shared-handle <resource>)
```
I know I said "shared handle" earlier, but I forgot that `shared` is already claimed to mean "can be called/accessed racily from multiple threads" (as in `(memory 1 shared)` and, in the future, `(func shared ...)`).  Instead of introducing a different adjective, I think it's fine to simply call the copyable version of handles `(handle <resource>)`, treating this as the base case, on which later `(unique <resource>)` and `(borrow <resource>)` refine.

lukewagner commented on CanonicalABI.md:
````diff
@@ -0,0 +1,976 @@
+            | (pull-buffer <intertype>)
+```
+
+This is also intended to cover the various shorthands in the `Explainer.md` as
+well (e.g. `string` is simply `list char`). This grammar adds three new types
+over what's explained in the explainer:
````
```suggestion
This is also intended to cover the various shorthands in the `Explainer.md` as
well (e.g. `string` is simply `list char`). This grammar includes three types that have
been tentatively discussed, but are not yet present in the explainer.  These types are
necessary to complete a fully interface-typed WASI and will be updated if they change
upon addition to the explainer.
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+  object a wasm module imports from a host. Handles are intended to be
+  un-forgeable capabilties where once a handle is given to a module it's seen as
+  granting a capability to a module, but the module cannot have access to
+  handles that it hasn't explicitly been granted access to.
```
```suggestion
  handles that it hasn't explicitly been granted access to.  Handles are described in more detail in
  [this presentation to the WASI subgroup](https://docs.google.com/presentation/d/1ikwS2Ps-KLXFofuS5VAs6Bn14q4LBEaxMjPfLj61UZE).
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+
+## Core wasm signatures
+
+First let's take a look at the core wasm signature for any particular interface
+types signature. Note that the core wasm signatures for interface
+types only use the `i32`, `i64`, `f32`, and `f64` wasm types. The `externref`,
+`funcref`, and `v128` types are not used in the core ABI.
```
```suggestion
First let's take a look at the canonical core wasm signature for any particular interface
types signature. Note that the canonical core wasm signatures for interface
types only use the wasm MVP types `i32`, `i64`, `f32`, and `f64` (not the post-MVP
types `externref`, `funcref`, and `v128`).
```

lukewagner reviewed (commented):
(Posting what I have so far, more in a bit. Mostly just wording nits.)

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+
+        # Records have a specialized representation when they're a
+        # struct-of-bools to represent bitflags. Otherwise it's the flattening
+        # of each member individually
```
I wonder if, to avoid the canonical ABI feeling arbitrary in what it packs, we should either:
1. adding a "TODO:" to consider more general packing optimizations (of which "all bools" is just one case); or
2. always flatten records (unpacked), and instead add the bitfield optimization for the `flags` abbreviation (and perhaps generalize it to work for any number of flags (thus producing either 1 `i32` or N `i64`s)).

alexcrichton replied:
It seems reasonable to me to not special-case booleans here and go ahead and do packing otherwise. I'm assuming though the packing would only happen for structs-of-enums where the number of bits to represent all the enums is less than 64?

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+    variant $cases =>
+        s = 0
+        for case in $cases
+            s = max(s, size(direction, record([i32, case])))
```
Variant case payloads are optional in the grammar, in which case I think it'd be great of the payload didn't consume any bytes, meaning that, e.g., the `enum` abbreviation (with all empty cases) was always a single `i32`.

alexcrichton replied:
Ah yeah that's the intention, I think I just didn't model it  well. I'll update all references here and there to explicitly show that each `case` is a nullable thing

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+    variant $cases =
+        a = 4
+        for case in $cases
+            a = max(align, align(case))
```
```suggestion
            a = max(a, align(direction, case))
```

lukewagner commented on CanonicalABI.md:
````diff
@@ -0,0 +1,976 @@
+relatively straightforward:
+
+```
+def read_from_memory(direction, type, ptr):
````
Just for consistency, how about calling it `deserialize_from_memory`, so it's a clear variation on `deserialize`?

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+          # otherwise `read_from_memory` is defined below.
+          result = []
+          for i in 0..len:
+              read_from_memory(direction, $ty, ptr)
```
```suggestion
              result.push(read_from_memory(direction, $ty, ptr))
```

lukewagner commented on CanonicalABI.md:
````diff
@@ -0,0 +1,976 @@
+
+```
+
+## Lifting values
+
+With the core wasm signature above, the definition of how to lift values in
+import/export contexts then looks like:
+
+```
+def lift_all(direction, types, values):
+    values = values.iter()
+    types.map(|ty| lift(direction, ty, values))
+
+def lift(direction, type, values):
````
Ok, so I finally realized that you're using `lift` in sortof the opposite direction meant by the lifting and lowering adapter instructions (you're not going from core values into interface values but, rather, from core values to source-language values).  So in the alternative terminology we discussed, I guess this section would be "Deserializing the canonical ABI into source-language values" and the function name would be `deserialize`?  It'd be good to explain this in more detail in the intro text.

lukewagner replied:
(oh, as a reminder, don't forget to search comments for "lift" and "lower")

alexcrichton replied:
Yeah I like the renaming and I'll try to call this out in the intro text more.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+    u64 => values.assert_next(i64) as u64
+
+    # This assumes that the i32 is a valid unicode code point.
+    char => values.assert_next(i32) as char
```
```suggestion
    # We can assume that the i32 is a valid unicode scalar value
    char => {
      codept = values.assert_next(i32) as u32
      assert(codept < 0xD800 || (codpt >= 0xE000 && codept <= 0x10FFFF))
      codept as char
    }
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+      # call the `{$resource}_drop` imported (e.g.  # "fd_drop" for WASI).
+      if direction == "import":
+        assert(!$resource.owned_by_current_module);
+        value as $foreign_handle<$resource>
```
For lists, you use `defer` as your way of saying "some time later when you're done with it, free this list", could you do that here too for `{$resource}_drop`, moving some of the comments into code?

alexcrichton replied:
I updated the previous section to avoid the use of `defer` to just call it at the end of the loop, and I added some wording here about how the returned value here is expected to have a destructor but otherwise I don't think `defer` is right in this location since the destructor is tied to the lifetime of the `$foreign_handle<$resource>` value.

I may be getting too confusing with all the made-up syntax and such used throughout here...

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+        # free function of some form. For now that function is called
+        # "witx_free" and takes pointer/size/alignment arguments
+        if direction == "import":
+          defer module.witx_free(ptr, len * sizeof(direction, $ty), align(direction, $ty))
```
```suggestion
          defer module.witx_free(ptr, len * size(direction, $ty), align(direction, $ty))
```
Also, how about `canonical_free`, to match the imported builtins?

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+      # is the the same as the "import" case above where we wrap it up and
+      # when we're done with it we must call the "drop" function.
+      else:
+        value as $foreign_handle<$resource>
```
Also `defer` here?

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+      assert(direction == "export")
+
+      # This conversion is relatively simple, we're just wrapping up the
+      # `handle` in a language-specific value. More on this is defined below.
```
Maybe note how we don't need the `defer` b/c of the scoped semantics.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+
+        if $ty == char:
+          # note that this can skip validation due to the canonical ABI
+          read_utf8_string(ptr, len)
```
An interesting question is whether we might want to have the canonical ABI treat `string` differently than `(list char)` (as, in general, any binding is allowed to do for abbreviations (e.g., JS API producing a JS string vs. array-of-single-char-strings), with `string` decoded as UTF-8 and `(list char)` decoded according to `read_from_memory` which means, in effect, UTF-32.  I don't know if it's a good idea, but... I thought I'd ask.

alexcrichton replied:
Perhaps yeah, I was assuming that the abbreviations like `string` aren't actually present in the type grammar and they're basically just a parsing concern. I could see, though, having separate types `string` and `(list char)` which are compatible types but represented differently. I don't feel too strongly one way or another though.

lukewagner commented on CanonicalABI.md:
````diff
@@ -0,0 +1,976 @@
+    }
+```
+
+## Lowering values
+
+Lowering is very similar to lifting, it's basically just the opposite in various
+directions.
+
+```
+def lower_all(direction, types, values):
+    values = values.iter()
+    types.map(|ty| lower(direction, ty, values))
+
+def lower(direction, type, value):
````
... and similarly here I'll assume this would be "Serializing language values into the canonical ABI" and `serialize`?

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+        # here.
+        if is_bitflags($fields):
+          if $fields.len() > 32:
+              values.assert_next(i64) as $bitflags<$fields>
+          else:
+              values.assert_next(i32) as $bitflags<$fields>
```
Here's another interesting argument for treating `(flags "a" "b")` different from `(record (field "a" bool) (field "b" bool))`: if we're simply optimizing the canonical ABI of certain records, then the produced language value should be a `$language_struct` (as below).  But if we're recognizing the `flags` abbreviation, then we could produce a language-visibly-different value (e.g., an actual bitvector).

alexcrichton replied:
This is related to the `string` thing above where I've been working under the assumption that the syntactic sugar isn't present in the AST. In that sense the two records you described here wouldn't have a different representation in the source language, they'd always be bitflags.

I do think it's somewhat nice to not have too much special-casing and too complicated of an AST here for the type grammar, and I'm not sure if there's much use in having different canonical ABI representations for the same structs?

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+	  wasm_vals = []
+	  for field in $fields:
+	    wasm_vals.extend(lower(direction, field, value[field.name]))
+          return wasm_vals
```
```suggestion
         return wasm_vals
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+      match value {
+        the_case(payload)=> {
+          # Recursively lower the optional payload
+          payload_vals = lower(direction, $cases[the_case], payload)
```
Since it's optional, I think maybe this goes in an if (since `lower` takes a non-optional type)?

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+          for i in len(payload_vals)..len(payload_types)
+              payload_vals.push(zero_value(payload_types[i]))
+
+          [the_case as wasm_i32] + payload_vals
```
```suggestion
          [$cases.index(the_case) as wasm_i32] + payload_vals
```
(My understanding is that `the_case` is a string label, and thus we should be explicit that the integer ordering comes from what is declared in the (local) type.)

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+handle. The reference count increases as the resource crosses module boundaries,
+and the reference count decreases when a module willingly gives up the handle to
+the resource (by calling the drop function).
+
```
I think we need two more intrinsics so that wasm is able to virtualize interfaces with handles:
* an intrinsic for creating a new handle (implemented by the caller) in which the caller passes an `i32` (probably a pointer into linear memory)
* an intrinsic for getting back the `i32` for a handle, but only if that handle is `owned_by_current_module` (trapping otherwise)

alexcrichton replied:
Oh this was intended to be covered automatically in serialization/deserialization/adaptation. Creating a handle is simply passing the i32 out to another module and re-acquiring the i32 happens when the handle is passed back to you automatically.

I'm not sure why other intrinsics would be needed though, so you may need to walk me through what you're thinking here

lukewagner replied:
Ah hah, I see.  Yes, that seems simpler.  I wonder if it loses the ability to express any interesting use cases, though?  E.g., let's say my component wants to create a handle once and hand it out N times.  This could ensure that my destructor was only called once (when the last of the N handles was done).  Otherwise, I'd need to create N handles, each of which is separately refcounted, and then implement a second refcount in linear memory (the N handles' destructors calling decref), which may end up introducing an extra level of indirection.  In the general (custom adapter function) [sketch](https://docs.google.com/presentation/d/1ikwS2Ps-KLXFofuS5VAs6Bn14q4LBEaxMjPfLj61UZE/edit#slide=id.gb925422856_0_34), there is an explicit `resource.new` instruction for creating a resource (padding in the `i32` payload), and getting back a handle to it, and a separate `resource.get` instruction for getting the `i32` payload out, so these two builtins I mention would essentially correspond to those.  WDYT?

alexcrichton replied:
Nah that all makes sense to me. I'll switch to an intrinsic like this.

alexcrichton replied:
Hm ok actually thinking about this I'm less certain that this is a good idea. This I think muddies the ownership story and makes for some unnecessary reference counting.

For example if a module wants to return a handle to something it just created, it'd first call an intrinsic to wrap up a private 32-bit value into a handle. This creates a handle with reference count 1. Ideally the module would transfer ownership of that handle to the outside world, but that's not how handles are specified right now in lifting/lowering (that requires the unique/borrowed/etc). This means that the callee now has a strong reference count of 1 which it didn't have anywhere to store since it just returned it. It's not clear when the decrement should happen?

Similarly though if a module receives a handle from the outside world, what it ends up receiving is a fresh index from the outside world for a new handle given the existing lifting/lowering. It would then call an intrinsic to acquire the internal private 32-bit value on this handle. After the call finishes it would have to then deallocate this owned handle.

Overall I think what you're saying here makes sense, but I think it only makes sense in a world where there's shared/borrowed/unique handles. The current lift/lower for handles I've documented effectively assumes that returning handles is always returning a unique handle, and taking a handle always takes a borrowed handle. I think what you're spec'ing would be "what if both cases were shared handles".

Typing this all out I realize that handles are intended to be shared handles today, but this feels like a lot of overhead for something that probably won't be desired that much. There's overhead of extra calls in the "receive a handle" case, and in the "return a handle" case I'm not actually sure where the decrement of the reference count would generally go...

lukewagner replied:
In some sense, you've already implemented a shared handle, since sharing is what happens when A passes a handle to B, with neither A nor B being the owner.  What I'm proposing is that things work the same way even when A is the resource owner (b/c A will have already created the resource, sticking a Handle into A's own TABLE).  I *think* this answers the ref-counting questions?

You're right that there is unwanted extra builtin-call overhead, though.  One idea is to say that the canonical ABI for lifting a handle (when you are the resource type owner) is to pass two `i32`s: the first one says "create a new one, the second `i32` is the private value", the second one says "i already created the handle, the second `i32` is the index in my handle table".  That means that the canonical ABI for a handle type depends on the resource type, but I think that's fine, since it's static.  WDYT?

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+      }
+    }
+
+    # It's expected that handles are always represented in a language as a
+    # 32-bit integer, regardless of whether it's defined in the language itself
+    # or a reference to a foreign handle.
```
```suggestion
    # It's expected that handles are always represented in a language as a
    # 32-bit integer (which logically represents an index into a handle table
    # maintained by the canonical ABI), regardless of whether it's defined in 
    # the language itself or a reference to a foreign handle.
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+
+    # Buffers being lowered only happens for the parameters of imports, in which
+    # case we're passing a either a pointer/length relative to our own module or
+    a buffer that was passed to us from somewhere else. This is similar to the
```
```suggestion
    # a buffer that was passed to us from somewhere else. This is similar to the
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+    push-buffer $t | pull-buffer $t => {
+      assert(direction == "import")
+      if value.is_local_buffer:
+        [0 as wasm_i32, value.ptr as wasm_i32, value.len as wasm_32]
+      else:
+        [1 as wasm_i32, value.handle as wasm_i32, 0 as wasm_32]
```
How about the idea discussed at the end of the last chat where:
* 0 is guaranteed to be an invalid handle index
* the first `i32` of a local buffer is given `0`
* the first `i32` of a non-local buffer is given the (non-zero) `value.handle`
* non-local buffers also have a `value.view_offset` and `value.view_len` which are normally 0 and the length of the buffer, but can be reduced via some language-specific `slice()`

lukewagner reviewed (commented):
Here's the next batch of comments.  Overall, fairly minor nits; looking good!

lukewagner commented on CanonicalABI.md:
````diff
@@ -0,0 +1,976 @@
+
+```
+# This is the pseudo-code that a host might execute when it's calling an
+# exported function from a wasm module. In this case `params` is of the shape
````
```suggestion
# This is the pseudo-code that a host might execute when it's calling the
# exported function of a wasm module. In this case `params` is of the shape
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+# exported function from a wasm module. In this case `params` is of the shape
+# of whatever language the host is written in.
+def host_call_wasm_export(signature, params):
+    wasm_values = lower_all("export", signature.params, params)
```
IIUC, `lower_all`/`serialize_all` will pass a (host) `list` will try to pass the host's list pointer as an `i32`.  It seems like some sort of `malloc` has to happen here.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+      lift_params = wasm_params
+    values = lift_all("import", signature.params, lift_params)
+    results = call_the_host_import_implementation(..values)
+    wasm_results = lower_all("import", signature.results, results)
```
... and similarly here I think you need a variant of `serialize_all` that mallocs+copies out of host memory.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+copies and validating in-place and/or using in-place pointers. Assuming that a
+host is "just another instance" makes the definitions below a bit easier.
+
+Finally it's worth pointing out that the pseudo-code below happens when to
```
```suggestion
Finally it's worth pointing out that the pseudo-code below happens when two
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+the pointer is an array of 64-bit elements and each individual result is
+read/written to each slot in the array.
+
+## Adaptation of imports and exports
```
Maybe rename the section to "Adapter Fusion" and say something like "This implements the equivalent of [adapter fusion](https://github.com/WebAssembly/interface-types/blob/master/proposals/interface-types/Explainer.md#adapter-fusion) when all adapters are canonical."

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,976 @@
+    # It's the responsibility of the native language to supply a list where the
+    # in-memory representation of the list matches the canonical ABI.
+    list $ty => {
+      [value.ptr as wasm_i32, value.len as wasm_i32]
```
If this case is supposed to define the canonical ABI for lists, this this seems like a circular definition.  I suppose you could say that the `list` case in `deserialize` defines the canonical ABI, but only in a sortof indirect way.

lukewagner reviewed (commented):
My comment for `list` in `serialize` makes me wonder if the right way to define the canonical ABI in this doc is not to specify `serialize`/`deserialize` (which is how you go between the canonical ABI and source language values) but, rather, `list`/`lower` (which is how you go between canonical ABI and *interface* values and is the reverse direction from how `lift`/`lower` are currently written).  It seems like it would make a number of things that are a little fuzzy in the current doc (due to the fact that we're not identify a single source language) more precise.

One other interesting consequence is that there wouldn't need to be a wholly separate `adapt` function: `adapt` could simply invoke `lift` and `lower`.  To achieve the interleaved side effects that we need, I think the Python pseudo-code for `lift` and `lower` would need to be generators: calling `lift(values)` would give you a generator that, every time you pumped it, gave you the next lifted value.  So `def fuse(sig, a_instance, b_instance, params): lower(sig.results, lift(sig.results, b_instance.exports[sig.name](lower(sig.params, lift(sig.params, values)))))`

alexcrichton commented:
notes to myself:

* generators are necessary for order of function calls when wasm-to-wasm is fused
* host interactions can probably be eschewed by basically saying "it should be relatively obvious how to convert your host values into the stuff that interface types is using". It's roughly along the lines of using `lift` and `lower` except you probably want some optimizations in some places.
* Add new type variants for `string` and `flags` so `list(char) != string`

alexcrichton commented:
Ok I think this turned out much better this time around with lifting/lowering/fusion. I think it still gets the point across in a straightforward way how code generators might implement it and I think it's still clear on what's happening where.

I think I've taken all your comments into account (except one unresolved one above about intrinsics for handles). I've tried to fill in more information as well on handles and buffers, notably defining reference counting semantics for handles and the `push`/`pull` imported functions for buffers and how they're supposed to operate.

alexcrichton commented:
Oh one thing I am still missing though is the switch to using generators. I couldn't really quite see where things would fit in, especially with the current definition of `fuse` where it does the interleaving already.

I wasn't sure as well if you were thinking we'd do the sax thing or whether we'd yield more first-class values. Overall though I thought that what's currently here roughly represents the desired semantics, so I wasn't sure where generators would slot in, but you'd know for sure!

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+            | char
+            | string
+            | (list <intertype>)
+            | (bitflags <name>*)
```
Currently, `bitflags` is just `flags` in the Explainer; is there a reason for changing that?

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+  (e.g. `string` is utf-8 while `list char` is utf-32). This aspect of the
+  proposal may change over time though as the types are specialized in a
+  different way, but regardless the ABI is expected to be different than the
+  otherwise equivalent structure.
```
```suggestion
* `string` and `flags` - while the explainer currently defines these as specializations
  of `(list char)` and `record`-of-`bool`, resp., specializations are specifically allowed to
  have different representations in bindings (such as the canonical ABI) than the general
  type they specialize (e.g. `string` is utf-8 while `list char` is utf-32).
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+    # bitflags are represented as a contiguous list of `i32` values
+    bitflags $fields => {
+      fields = []
+      for _ in 0..align_to($fields.len(), 32):
```
If I'm not confusing myself, for the purpose of bit-packing, you want `align_to($fields.len(), 32)/32`; otherwise all your logic here and below is off by a factor of 32.  How about a new function `num_packed_i32s($fields)` that you define here (which also lets you spell out clearly the 0 case).

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+`pull-buffer` are represented with a singular `i32` instead of three `i32`
+values. This is because if an export is receiving a buffer it's coming from
+somewhere else which means it's a handle to a buffer and various "intrinsics"
+will be used to operate on the handle.
```
```suggestion
somewhere else which means it's a handle to a buffer and various "intrinsics",
listed below, will be used to operate on the handle.
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+The `lower` function here takes three parameters:
+
+* `direction` - one of `"export"` or `"import"` which indicates for which
+  purposes the lowering is happening.
```
```suggestion
* `direction` - one of `"export"` or `"import"` which indicates for which
  purposes the lowering is happening: lowering an export means lowering
  a parameter, lowering an import means lowering a result.
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+    char(val) => [val as wasm_i32]
+
+    string(val) => {
+      base = dst.canonical_abi_malloc(val.len(), 1)
```
```suggestion
      base = dst.canonical_abi_malloc(utf8_byte_length(val), 1)
```

lukewagner replied:
I realized this should probably be a `dst.canonical_abi_realloc`.  The (subtle) reason is this: while the canonical ABI will always have the `utf8_byte_length` "on hand", with custom adapter functions (in the general case), this won't be knowable ahead of time.  Instead, I think what the canonical ABI needs to do is use a dynamic `realloc`-based scheme.  Because canonical-ABI-components will be used with future-custom-ABI components, we need to prepare for this by requiring canonical-ABI components to give us a full-fledge `realloc`.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+      for item in val:
+        lower_to_memory(direction, dst, ptr, item)
+        ptr += size(direction, $ty)
+      [base as wasm_i32, len as wasm_i32]
```
```suggestion
      [base as wasm_i32, val.len() as wasm_i32]
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+    list<$ty>(val) => {
+      base = dst.canonical_abi_malloc(val.len() * size(direction, $ty), align(direction, $ty))
+      assert_or_trap(base % align(direction, $ty) == 0)
+      len = val.len()
```
```suggestion
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+    bitflags<$fields>(val) => {
+      values = []
+      for i in 0..align_to($fields.len(), 32):
+        values.push(val.bits & (0xffffffff << (i * 32)))
```
To support subtyping (in which superfluous fields/flags can big ignored), I think you need to iterate over `$fields` (as with the `record` case), updating `lift` to produce a dictionary (not bitfield)

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+    }
+
+    bitflags<$fields>(val) => {
+      for i, bits in lower(direction, dst, val):
```
```suggestion
      for i, bits in enumerate(lower(direction, dst, val)):
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+    variant<$cases>(val) => {
+      match val {
+        $case(payload) => {
+          # Write the discriminant, using the smallest size available
```
I think `align` and `size` would need to be updated for this to result in an optimization.  Is there a particular case you're thinking about optimizing here?  Otherwise, I'm wondering whether it's worth including this optimization.

alexcrichton replied:
The main thing I was trying to do with this was to match what WASI currently does which is roughly to shrink enum discriminants to the size  of the discriminant space. I figured that if WASI is doing something already it'd be nice to keep it in the canonical ABI to avoid clashes with "but WASI was faster?"

lukewagner replied:
Maybe we can forgo this optimization and add it back if anyone seems to especially care?  Otherwise, I think you'll also need to update `align` and `size`.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+The `lift` function here takes four parameters:
+
+* `direction` - one of `"export"` or `"import"` which indicates for which
+  purposes the lifting is happening.
```
```suggestion
* `direction` - one of `"export"` or `"import"` which indicates for which
  purposes the lifting is happening: lifting an export means lifting
  a result, lifting an import means lifting a result.
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+          # otherwise "downcast" all payload values to the type for this case
+          my_types = flatten(direction, payload)
+          payload_values.truncate(my_types.len())
+          for i, ty in my_types:
```
```suggestion
          for i, ty in enumerate(my_types):
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+      bits = 0
+      for i in 0..align_to($fields.len(), 32):
+        bits = (bits << 32) | read_4_bytes(src, ptr + i * 4)
+      bits as bitflags<$fields>
```
Perhaps read out N `i32` values and delegate to `lift`, as you did with `lower_to_memory`, which also lets you pick up the dictionary change suggested above.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+    variant<$cases> => {
+      match val {
+        $case(payload) => {
+          # read the discriminant, using the smallest size available
```
(same comment as above)

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+lifting/lowering will work for managing indices. Note that the implementation
+provided here is not canonical. The indices used to refer to handles and buffers
+can vary between engines and their values should not be relied upon. A sample
+implementation, however, is provided here to show some of the desired semantics.
```
Actually, in pursuance of our general goal of determinism in wasm, I think the indices *should* be semantically-defined (for the same reason that, e.g., you zero out extra variant args).  Otherwise, you could have some bizarre bug that only manifests when, say, two indices hash to the same value and it would only show up in some implementations.

alexcrichton replied:
Hm so do you think that we should canonicalize the index allocation algorithm? That's the only other option I think?

lukewagner replied:
Yep, I think that's what we have to do.  It just means strengthening the comments on your `TABLE` code to say "while the impl can be anything, the issued indices have to be exactly what this psuedo code would do".

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+  #
+  # Returns a 32-bit integer which can later be passed to `get` to retrieve the
+  # original `handle`. This function always succeeds, unless the table
+  # overflows from too many insertions.
```
I suppose we should define the overflow case to return -1 (as with `memory.grow`) so that modules can be OOM-safe, if they so choose.

alexcrichton replied:
Thinking more on this and the `realloc`, this raises a few questions for me:

* How come we're trying to future proof against future more general adapters? In that world wouldn't you be able to specify whatever adapter code you wanted and nothing would be baked in? For example this `TABLE` would be entirely implemented in adapter code (or by engines with `resource.new` and such), so if this is purely an implementation of the canonical ABI, and we're assigning just one set of semantics to the canonical ABI, I'm not sure why we'd need to future proof so modules can do something on their own.
* Do your concerns here also apply to `malloc` allocation failure? I've currently documented that `canonical_abi_malloc` is expected to trap if allocation fails, but if we're going the route of future-proofing should it also return -1?

Although I suppose that this could also be interpreted in the context of your other comment where the TABLE bits here should be intrinsics, not implicitly happening as part of the glue. That would mean that modules themselves would receive the -1 and decide what to do, which in that case I would agree is not future proofing it's just letting modules be more flexible.

(my question about future-proofing though would still apply to malloc/realloc where those are still only called by adapters of the canonical abi)

lukewagner replied:
The interaction I think we have to future-proof for is when a custom-adapted component calls a canonically-adapted component.  If the canonically-adapted component effectively requires the list size to be known up-front (so it can issue a single `malloc` call), but the custom adapter is doing, say, UTF-16 decoding (where the number-of-USVs isn't known until the end), then that forces the glue to buffer up the decoded `char`s (an intermediate copy) in a first USV-counting pass, then `malloc`, then copy the `char`s into the malloced buffer.

Incidentally, I think we need not just `realloc` to be defined, but to actually used in practice (in place of `malloc`).  I still worry we'll get components with broken `realloc`s that only show up with custom adapters...

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+  new_params = []
+  for param_ty in signature.params:
+    val = lift("import", instance_a, param_ty, params_to_adapt)
+    new_params.extend(lower("export", instance_b, val))
```
Perhaps add an assertion that `params_to_adapt` is empty (ensured by validation).  If `lift` returned a generator, you could assert it was done as well.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+      ptr = values.assert_next(wasm_i32)
+      len = values.assert_next(wasm_i32)
+
+      val = read_utf8_string(src, ptr, len)
```
It might be good to comment that `read_utf8_string` needs to validate UTF-8 since it can't trust it's valid.  I've been thinking that, rather than trapping, `read_utf8_string` should precisely follow the [TextDecoder.decode algorithm](https://encoding.spec.whatwg.org/#dom-textdecoder-decode) with the [replacement error mode](https://encoding.spec.whatwg.org/#textdecoder-error-mode).  Instead of trapping, this algorithm will produce replacement characters, which I think is what you want if you have a WTF-16 language like JS/Java/C#.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+
+    string(val) => {
+      base = dst.canonical_abi_malloc(val.len(), 1)
+      encode_utf8_string(dst, base, val)
```
It's probably worth a comment noting that we are guaranteed only valid incoming [unicode scalar values](https://unicode.org/glossary/#unicode_scalar_value).

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+Here `fuse` performs the necessary lifting and lowering to translate all
+parameters and results across the call. This `fuse` function is expected to be
+implemented by "trusted code" since it refers to `TABLE` and otherwise performs
+validation of values (e.g. that a `string` is indeed valid UTF-8).
```
```suggestion
implemented by "trusted code" since it refers to `TABLE`.
```
(Now `lift` does validation.)

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1082 @@
+    # the host can retain the raw pointers into the wasm module's memory if it
+    # likes.
+    list<$ty>(val) => {
+      base = dst.canonical_abi_malloc(val.len() * size(direction, $ty), align(direction, $ty))
```
ditto for `realloc`

alexcrichton replied:
I included `realloc` in the section about intrinsics, but do you think it also needs to be mentioned around the lifting/lowering?

lukewagner reviewed (commented):
This is looking great, I really like the new lifting/lowering!

So for the generator change: I was realizing it could be simpler than I said earlier: instead of converting the nested types into a "token stream" (with "start list" and "end list" tokens, etc), I think you can:
* for records, just yield a sequence of (`field`, `value`) pairs
* for variants, first yield the discriminant (string label) then yield the payload sequence
* for lists, yield a generator that yields the elements of the list

I think this should allow you to keep basically the same function structure as you have now and just make somewhat superficial tweaks.  And again, the point of this change would be to accurately describe the order of side effects within `fuse` (in particular, the interleaving of `lift` and `lower`).

alexcrichton commented:
Ok and the latest commit is my attempt to use generators, can you double-check me though?

I don't think there's necessarily an obvious place to handle subtyping and stuff just yet, but it probably shouldn't be too hard to add in the future?

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1137 @@
+
+* `direction` - one of `"export"` or `"import"` which indicates for which
+  purposes the lifting is happening: lifting an export means lifting
+  a result, lifting an import means lifting a result.
```
```suggestion
  purposes the lifting is happening: lifting an export means lifting
  a result, lifting an import means lifting a parameter.
```
heh, I had a typo in my last suggestion :P

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1137 @@
+      ptr = values.assert_next(wasm_i32)
+      len = values.assert_next(wasm_i32)
+
+      assert_or_trap(ptr % align(direction, $ty) == 0)
```
I think it will be better for optimization and also match the semantics of `memory.copy` (which fusion wants to emit) to have an `assert_or_trap` that `ptr+len` is in-bounds (before the first byte is read).  (Also in the `string` case.)

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1137 @@
+      yield(list_start<$ty>(len))
+
+      for i in 0..len:
+        yield(lift_from_memory(direction, src, $ty, ptr + i * size(direction, $ty)))
```
```suggestion
        yield from lift_from_memory(direction, src, $ty, ptr + i * size(direction, $ty))
```
If we're writing Python here, iiuc, `yield` will yield a generator object; `yield from` will do what you want here which is to yield all the recursive yields of `lift_from_memory`.

alexcrichton replied:
My lack of actual python knowledge is showing here...

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1137 @@
+    record<$fields> => {
+      yield(record_start<$fields>)
+      for $field in $fields:
+        lift(direction, src, values)
```
```suggestion
        yield from lift(direction, src, values)
```

lukewagner replied:
And, iiuc, you need to have all other recursive calls to `lift(...)`/`lift_from_memory(...)` be `yield from` to get the desired transitive yield behavior.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1137 @@
+            payload_values.truncate(my_types.len())
+            for i, ty in enumerate(my_types):
+              payload_values[i] = downcast(payload_values[i], ty)
+            lift(direction, src, payload, payload_values)
```
```suggestion
            yield from lift(direction, src, payload, payload_values)
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1137 @@
+      yield(record_start<$fields>)
+      for $field in $fields:
+        ptr = align_to(ptr, align(direction, $field))
+        yield(lift_from_memory(direction, src, ptr, $field))
```
```suggestion
        yield from lift_from_memory(direction, src, ptr, $field)
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1137 @@
+    # likes.
+    list_start<$ty>(len) => {
+      base = dst.canonical_abi_malloc(len * size(direction, $ty), align(direction, $ty))
+      assert_or_trap(base % align(direction, $ty) == 0)
```
Similar for lifting, it'd probably be good to add an `assert_or_trap` for in-bounds before doing any effectful copying.

lukewagner reviewed (commented):
Awesome, the generator change looks awesome.

alexcrichton commented:
Ok I'm sort of losing track of discussions due to so many GitHub buttons at this point. I'm gonna try to summarize those here:

* [Intrinsics for handles](https://github.com/alexcrichton/interface-types/pull/1#discussion_r610257383) - I'm somewhat wary to do the buffer thing for handles where the representation in the ABI has an extra `i32` for discriminant purposes. Not really for any concrete reason, mostly just because it seems like handles would be used a lot and an alternate form wouldn't be used all that much. My main worry is modules returning a handle that don't want to store their own module's copy of the handle. The only way to do that is to have the glue allocate a handle index in the destination module directly, because otherwise no one's around to deallocate the index in the original module. 
 
  I think on this though there may be some confusion? Your comment mentioned A passing a handle to B where neither is the owner, but in that case everything being passed around is table indices. Only when A passes to B where A is the owner (e.g. via returning from a function) does A's value get interpreted as being the private value to get wrapped up. I may be a bit confused though, I could try writing more of an example?

* [size of variant discriminants](https://github.com/alexcrichton/interface-types/pull/1#discussion_r617902072) - one of my main concerns here was `bool`, for example, which is 32-bits or 8-bits depending on this optimization. All-in-all I think it's pretty small and seems reasonable enough to implement? Once nice benefit of shrinking the discriminant is it's probably more likely to match the host representation in more cases by default, but that's not a super strong argument. I'm mostly trying to head off a switch for `wasi-libc` to use the canonical ABI and folks being surprised that booleans are 32-bits in memory instead of 8.

* [determinisim of table indices](https://github.com/alexcrichton/interface-types/pull/1#discussion_r617929877) - I don't mind specifying the slab-style allocator, but I thought one of the goals of WASI was to allow engines to have the freedom to allow randomized index allocation? Or is this something where the glue between modules should be canonical but the host can do whatever it wants? I'm just a bit wary to give up some flexibility when no one will really be able to take advantage of the flexibility I think.

* [`realloc` vs `malloc`](https://github.com/alexcrichton/interface-types/pull/1#discussion_r617930997) - for this are you thinking that we just don't have `malloc` in the canonical ABI? But instead a fresh malloc is something like where -1 is passed as the parameter to the "previous allocation"? (this seems reasonable to me just want to clarify)

lukewagner commented:
Thanks so much for collecting all the issues in one place and summarizing!
* `Intrinsics for handles`: Those are great points, esp. "My main worry is modules returning a handle that don't want to store their own module's copy of the handle."  I suppose we can ust say that the canonical ABI doesn't have to represent 100% of use cases -- custom adapters can fill that gap in the future.  So agreed it's good as is.

  Independently, I'd like to dig into the confusion you're describing in your second para.  My reading of the lifting/lowering code is that, when A passes a handle to B, neither being the owner, A will `Table.get()` with A's index to get a Handle, the Handle will be given to B, then B will `TABLE.insert()` the Handle, getting a new index for B.  And this is necessary since each instance (A and B) logically has its own private, encapsulated table.  Thus, I think instances should never just be passing indices between each other (only to their respective canonical adapter functions)?
* `size of variant discriminants`: Ok, I think you're right.  I was going to propose doing something special for the `bool` specialization, but then why not `enum`s and then we're most of the way to `variant`...
* `determinism of table indices`: From a spec perspective, the canonical adapters behave as-if they were just custom adapter code executing *inside* the component, and thus it's not really the "host" that is choosing the particular index values -- it's the adapter code (the host is semantically receiving/producing an opaque handle value).  Moreover, since adapter code is determinisic, we'd have to do something magic (i.e., not simple desugaring) to allow for random indices.  Overall, I think the general WASM/WASI theme is determinism, so this is a good default for us here.
* `realloc vs. malloc`: Yup!  [From what I read](https://pubs.opengroup.org/onlinepubs/009604599/functions/realloc.html), `realloc(NULL, size)` behaves as `malloc(size)`, so I think we should use `0` instead of `-1` to line up with this libc convention.

alexcrichton commented:
Ok I've marked the discriminant bit resolved and I've switched to `realloc` and dropped `malloc`, marking that bit resolved.

* For more info on intrinsics for handles, you're right on A passing to B with neither is the owner. When I say "everyone passes table indices" I'm glossing over the fact that the table index gets rewritten in the middle by the lifting/lowering. My worry thought was in the case that A owns a handle and returns it, so the case where either was the owner wasn't quite as applicable.
  
  One possible saving grace is the we do have the ability to update the canonical ABI in the future through separate types perhaps? We could add something like `handle2` which typechecks and validates the same way as `handle` but simply has a different ABI of how it's passed? The alternative is that we could go ahead and add `unique` and `borrowed` handles and use those everywhere. The intrinsics you're describing would be used for `shared` handles and currently what I've specified is pices of the lifting/lowering for unique/borrowed handles.

* For determinism and table indices, I agree that whatever the module does should be deterministic and WASI's goal is also determinisim. I'm afraid of canonicalizing a half-baked algorithm though. For example this treats modules as having one table index space instead of one table index space for each type of handle in a module. Furthermore I do think we should remain flexible in handle values so if you wanted to globally configure "start all handles at N" or something like that it feels like it should be a valid configuration. I feel like the way it should be worded is that the lifting/lowering code should be deterministic but it's not necessarily the same values across implementations. 

  In some sense though this is only a problem for wasm->wasm calls, and there may be very few implementors of that beyond our own eventual `wasm-link` idea...

We could perhaps do a zoom later today if you want to talk about these bits more?

alexcrichton commented:
Ok all updated now!

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1166 @@
+
+* There is a separate index space per-instance and per-resource type. This means
+  that each instance's adaptation effectively has access to a set of tables for
+  each hadnle type, and indices are managed by allocating from these tables.
```
```suggestion
  each handle type, and indices are managed by allocating from these tables.
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1166 @@
+  each hadnle type, and indices are managed by allocating from these tables.
+
+* Indices to refer to resources are handled in a LIFO fashion. Indexes are
+  allocated starting from zero and increasing afterwards as more space is
```
```suggestion
  allocated starting from one (ensuring 0/`NULL` is always an invalid handle) and increasing afterwards as more space is
```

lukewagner commented on CanonicalABI.md:
````diff
@@ -0,0 +1,1166 @@
+the reference count. They are implemented as follows:
+
+```
+def resource_clone(src, $resource, val):
````
nit: might want to rename to match the builtin import string

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1166 @@
+  # is returned, otherwise if a check fails this raises a trap.
+  def get_buffer(self, idx, instance, ty) -> Buffer:
+    idx = idx - 1             # it's a precondition idx is never 0 when called here
+    assert_or_trap(idx < self.buffers.len())
```
```suggestion
    idx = idx - 1
    assert_or_trap(idx >= 0 && idx < self.buffers.len())
```
While buffer lifting doesn't allow 0 indices, iiuc, the builtin-import callers of `TABLE.get_buffer` don't check; they just take an index directly from the core wasm caller.

lukewagner commented on CanonicalABI.md:
````diff
@@ -0,0 +1,1166 @@
+
+## Memory Intrinsics
+
+Lists sometimes need to be communicated between modules and this necessitates
+the need to allocate and free memory in modules. If needed the realloc/free
+functions will be exposed as:
+
+```wasm
+(module
+  (func (export "canonical_abi_realloc") (param i32 i32 i32 i32) (result i32)
+    ;; ...
+  )
+  (func (export "canonical_abi_free") (param i32 i32 i32)
+    ;; ...
+  )
+)
+```
````
```suggestion
When lowering a list, the canonical ABI needs to allocate space in the destination
core module's linear memory that can be written into.  When lifting a returned list,
the canonical ABI needs to free the allocated memory after it has been read from.
When either of these are required, the core module (that's calling the canonical ABI)
must export one or both of the following functions:

```wasm
(module
  (func (export "canonical_abi_realloc") (param i32 i32 i32 i32) (result i32))
  (func (export "canonical_abi_free") (param i32 i32 i32))
)
```
```

lukewagner commented on CanonicalABI.md:
````diff
@@ -0,0 +1,1166 @@
+
+```
+class TABLE:
+  instances = Map<(Instance, Resource), Slab<Handle>>
````
It'd probably be good to implement Slab.  In particular, somewhere needs to clearly specify returning non-zero indices (as `insert_buffer` correctly does) since `push(handle)` is a bit unclear on this.

lukewagner commented on CanonicalABI.md:
````diff
@@ -0,0 +1,1166 @@
+  done so indices can fit into a hypothetical `i31ref` in the future.
+
+```
+class TABLE:
````
nit: maybe name this `TABLES` b/c it's *all* the tables, not just a particular instance+type's table.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1166 @@
+      ptr = values.assert_next(wasm_i32)
+      len = values.assert_next(wasm_i32)
+
+      assert_or_trap(ptr % align(direction, $ty) == 0)
```
I'm now wondering whether the canonical ABI should perhaps not `assert_or_trap` on alignment of linear memory allocations.  My reasoning is:
* `memory.copy` et al don't
* if the canonical ABI impl wants to call some optimized routine that requires alignment, the same branch needed for trapping could instead redirect to a byte-unaligned version.  (Although, practically speaking, I expect canonical ABI impls will just use `memcpy` to handle it for them.)

Thus, requiring alignment might just be making work for producer tools without providing much in the way of benefit for perf or engines.  Am I missing any other reasons?

alexcrichton replied:
The main downside which made me think that alignment should be respected is that all language bindings will need to use unaligned things I think? For example it's unsound in Rust to have `&T` where the memory is actually misaligned (due to LLVM optimizations), so I figured since everyone will respect alignment anyway we may as well go ahead and guarantee it in the ABI.

If we can solve that issue though it seems reasonable to not require alignment I think. Now that I think about it, because the module itself is always responsible for managing all the pointer (either they come from the source or from malloc), perhaps my concern here is a moot point? Rust can guarantee that everything is still correctly aligned but from the canonical abi perspective it's not inventing pointers out of thin air and it'll work with unaligned things if you pass it in.

I do think, though, we'd still want to pass the alignment to realloc/free, do you think that's ok?

lukewagner replied:
Yes, totally agreed.

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1166 @@
+  assert_or_trap(ptr + len * size("export", buffer.ty) <= src.memory.len)
+  amt = min(len, buffer.len)
+  for i in 0..amt:
+    val = lift_from_memory("export", src, buffer.ty, ptr)
+    lower_to_memory("export", buffer.src, buffer.offset, val)
```
```suggestion
    gen = lift_from_memory("export", src, buffer.ty, ptr)
    lower_to_memory("export", buffer.src, buffer.offset, gen)
```

lukewagner commented on CanonicalABI.md:
```diff
@@ -0,0 +1,1166 @@
+  assert_or_trap(ptr + len * size("import", buffer.ty) <= dst.memory.len)
+  amt = min(len, buffer.len)
+  for i in 0..amt:
+    val = lift_from_memory("import", buffer.src, buffer.ty, buffer.offset)
+    lower_to_memory("import", dst, ptr, val)
```
```suggestion
    gen = lift_from_memory("import", buffer.src, buffer.ty, buffer.offset)
    lower_to_memory("import", dst, ptr, gen)
```

lukewagner commented on CanonicalABI.md:
````diff
@@ -0,0 +1,1166 @@
+def resource_get(src, $resource, val):
+  return TABLE.get($resource, handle, src).val
+```
+
+These functions are used to create a handle wrapper around a resource
+(identified as a unique i32) and then to access the private i32 value. These
+functions are only available to modules/instances that defined the resource
+which means no other instances can have access to the private internals. Note
+that `get` will trap if the index passed in is invalid.
+
+Note that hosts may have their own reference count of resources owned by wasm as
+well. For example if a wasm resource makes its way all the way to a host then
+the host may hold a reference count that, when dropped, may trigger a wasm
+destructor. Similarly if a wasm resource is owned by the host and all other wasm
+modules with access drop the resource then the resource isn't dropped in the
+original module until the host is done with it.
+
+## Buffer Intrinsics
+
+Buffers defined in foreign modules are identified by an `i32` index. To actually
+operate on that index the following intrinsics can be used:
+
+```wasm
+(module
+  (import "canonical_abi" "push_buffer_len" (func (param i32) (result i32)))
+  (import "canonical_abi" "push_buffer_push" (func (param i32 i32 i32) (result i32)))
+  (import "canonical_abi" "pull_buffer_len" (func (param i32) (result i32)))
+  (import "canonical_abi" "pull_buffer_pull" (func (param i32 i32 i32) (result i32)))
+)
+```
+
+The implementation of these intrinsics is supplied by the table/buffer
+management glue.
+
+The `canonical_abi::push_buffer_len` function takes the integer descriptor for
+the buffer and returns the integer length of the buffer. This is the maximal
+number of items which can be pushed into the buffer.
````
Two related questions:
1. Does the buffer length decrease over time after repeated pushes/pulls to it (reflecting progress filling the buffer)?
2. Is it bad that a pull/push-buffer of `char` is effectively UTF-32-encoded?

For 2, I think it is a problem: folks wanting something more compact would be forced to use `(push-buffer u8)`, and now we have lots of potential for subtle encoding incompatibilities.  But if we use UTF-8 we have a problem: UTF-8 pull and push buffers wouldn't have an *a priori* number-of-units length.  (Esp. once you start considering custom adapter functions with arbitrary encodings on either side.)  I wonder if a solution is to change the push/pull len to be an "estimated" length remaining, which (answering question (1)) would decrease over time.  The idea being: the recipient of the buffer can use the estimate to push/pull about the right amount, and repeatedly pull/push until the estimate returned is 0.  One kindof neat thing is that, in the future with custom adapter functions, this could allow a custom adapter to use a single call and single push buffer to, say, fill a list-of-chunks.  Also, `char` might not be the only element type that wants a variable-length encoding with custom adapter functions.

WDYT?

alexcrichton replied:
For (1) yeah the buffer length is updated on each call to push/pull so subsequent calls only operate on what the remaining space in the buffer is. This is sorta in `def push` and `def pull` where they update `buffer.len` incrementally as items are moved back and forth.

For (2) you're right that `(push-buffer char)` is utf-32 right now. About the a-priori things, though, this feels a bit odd. You'd still want to construct a `char` buffer with ptr+len where len is the utf8-length, so on construction you don't know the number of `char` elements. This means that the adapter glue would have to implement utf-8 encoding/decoding to figure out the real number of `char` elements, but the `len` would be wrong for multibyte characters (always) unless the glue calculated that?

I do think it makes sense for the len to be "estimated" though in the long run. On construction though I think the `len` is required to always be exact because it's pointing in memory though?

lukewagner replied:
Right, so I was thinking that, on the canonical-lifting side, `len` does have to be the exact byte length.  Then, on the canonical lowering side, the estimate would be determined by the element type: for all types other than `char`, it would be `len/size(elem)`; for `char` it would be simply `len` (and thus wrong in the case of multi-byte, but that's ok, b/c it's just an estimate).

alexcrichton replied:
Oh currently on the lifting side `len` is an element length, not a byte length. If we say that `len` on the consuming side is just a hint, does that actually change much here? Perhaps renaming `len` to `size_hint` or something like that?

lukewagner replied:
Right, so I thinking different units for producer/consumer: producer provides number of bytes, consumer receives estimated number of elements (with the estimate being derived from the number of bytes)?

alexcrichton replied:
I originally avoided byte length to avoid dealing with the question "what if you give me 3 bytes for a struct that's 4 bytes large?" The obvious answer there is truncation but it seems like it would be best to not have to answer that at  all.

But I'm still not sure what is actually changing here? Bytes/elements would change the one division for calculation of the length, but otherwise are you thinking it's just documentation that's changing? (or maybe function names)?

Do you think there needs to be wording to say that `push-buffer char` is a utf-8 encoded buffer? (I'm not sure if it's really worth going to too much length here if that's not expected to be super common, which I'm not sure if it is?

lukewagner replied:
Ah, so are you thinking that, on the lifting size the given `i32` len could be "number of elements", except for `char`, where it would be "number of bytes"?  I suppose that could work too.

alexcrichton replied:
Oh I'm not thinking anything per-se but that could possibly work, yeah. I'm not personally convinced of the need to use utf-8 here so I'm trying to better understand how you (who I think wants to use utf-8?) would like things to work. In the end though I'm fine with whatever!

lukewagner replied:
Actually, I think I just talked myself out of it: if we wanted a UTF-8 buffer, by symmetry to `string`, we should do so by defining a buffer+string specialization (say, `*-string-buffer`).  Until we do that, by symmetry to `list char`, `*-buffer char` should be UTF-32.  So "n/m".

lukewagner reviewed (commented):
Awesome, really like these changes.  Scanning it all again from top to bottom helped me find one other interesting question a few more nits and suggestions.

lukewagner reviewed (approved):
Well I guess that's all my comments, thanks!
